### ISSUE TITLE
2.0.0 `everywhere` basically not working at all for complicated data structure

### ISSUE THREAD

**Issue by fommil** [milessabin/shapeless#320]
I tried to whittle my trait/case class domain model down to something minimal to show you but it basically didn't work at all when I had the model in a different file to the definition and application of the `everything`. I'd have to start again to show.

Is 2.1 going to have better support for this?

As a workaround, we're going to use runtime recursive Product traversal. If we can upgrade to scala-2.11 I'll try running shapeless 2.1 vs our runtime solution and let you know if anything breaks. If it does,  we should talk privately as I really can't share this domain model publicly and talking you through it might give you some ideas for a test case.

**Comment by fommil** [comment]
I've worked around this with runtime lookup. Maybe worth using something like this in your tests:

```scala
object ScalaVisitor {
  def everything[T <: Any](t: T)(f: Any => Unit): Unit = {
    f(t)
    t match {
      case t: AnyRef if isArray(t) =>
        t.asInstanceOf[Array[_]].foreach(everything(_)(f))
      case t: Traversable[_] =>
        t.foreach(everything(_)(f))
      case t: Product =>
        t.productIterator.foreach(everything(_)(f))
      case t =>
    }
  }

  def allTheThings(t: Any): List[Any] = {
    var seen = Queue.empty[Any]
    everything(t) { el => seen :+= el }
    seen.toList
  }

  def someOfTheThings[T](t: Any)(f: PartialFunction[Any, T]): List[T] =
    allTheThings(t).collect(f)
}
```

**Comment by milessabin** [comment]
I would guess that your original example involves (possibly hidden) recursion in a way which isn't supported in 2.0.0 ... can you boil you example down to something which you can compare between that and 2.1.0?

Other than that, I'm afraid that without something reproducible there's not a lot I can do apart from saying that what you're seeing isn't completely unexpected in 2.0.0 and is most likely fixed in 2.1.0.

**Comment by fommil** [comment]
I tried to boil it down, but it kept failing (because the classes were defined in a different file?) and I went too far, it is quite a big domain model so it would take me a few hours to try again. 2.1 sounds great, I'll just try and get us onto scala 2.11 and then I can report back with more meaningful results. No point trying to reproduce a bug on a branch that is known to be broken. On the plus side, this will actually be a reasonably good stress test of `everything`. Enjoy your holiday :smile:
